File: lc/go/binary-tree-paths.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type pair struct {
    node *TreeNode
    path []string
}
/*
            1
          /    \
         2      3
          \    /
           5  6
        ans: 125, 136
*/
func binaryTreePaths(root *TreeNode) []string {
    q := []pair {pair{root,[]string{strconv.Itoa(root.Val)}}}
    ans := []string {}
    for ; len(q) > 0; {
        pr := q[len(q)-1]
        q = q[:len(q)-1]
        root, rootpath := pr.node, pr.path
        if root.Left == nil && root.Right == nil {
            ans = append(ans, strings.Join(rootpath, "->"))
        } else {
            if root.Left != nil {
                leftpath := append([]string{}, rootpath...)
                copy(leftpath, rootpath)
                leftpath = append(leftpath, strconv.Itoa(root.Left.Val))
                lpr := pair {root.Left, leftpath}
                q = append(q, lpr)
            }
            if root.Right != nil {
                rightpath := append([]string{}, rootpath...)
                rightpath = append(rightpath, strconv.Itoa(root.Right.Val))
                rpr := pair {root.Right, rightpath}
                q = append(q, rpr)
            }
        }
    }
    return ans
}

File: lc/py/binary-tree-paths.py
from typing import List, Optional
# # Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
# when u get to the end put it in ans
# it: s -- keep appending/popping from s
# when hit leaf node add s to ans 

# base case: if none return 
# base case: leaf node: add urself, then add list to ans, then pop urself
# normal case: add urself, go

# ans []
# nodes [1 2 5]
# root 2
class Solution:
    def helper(self, nodes: List[int], root: Optional[TreeNode], ans: List[str]):
        if not root:
            return
        if not root.left and not root.right:
            nodes.append(root.val)
            ans.append("->".join(str(v) for v in nodes))
            nodes.pop()
            return
        nodes.append(root.val)
        self.helper(nodes, root.left, ans)
        self.helper(nodes, root.right, ans)
        nodes.pop()
        

    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        nodes, ans = [], []
        self.helper(nodes, root, ans)
        return ans
